tests/test_agents/base.py

```diff
@@ -252,34 +245,35 @@ def prepare(self, nb_nodes: int) -> None:
             self.__prepare_agent_i(agent_id, nb_nodes)
 
         # run 'aea install' in only one AEA project, to save time
-        self.set_agent_context(self.__get_agent_name(0, nb_nodes))
+        self.set_agent_context(self.__get_agent_name(0))
         self.run_install()
 
     def prepare_and_launch(self, nb_nodes: int) -> None:
         """Prepare and launch the agents."""
-        self.processes = []
+        self.processes = dict.fromkeys(range(nb_nodes))
```

Comment:
I refactored this such that processes are now a mapping and we don't allow for negative indexing anymore. This will ensure that whatever order you decide to terminate agents, and whatever other order you decide to restart them, the processes should be in the correct order. 

To illustrate the issue, terminating multiple agents and subsequently restarting them, either the one that terminated first will need to be restarted last, or the order of self.processes will be mixed up - e.g. when agent 4 terminates, then 3, then 4 restarts, then 3, we end up with processes [0, 1, 3, 2] because of appending to the list. Also popping first agents first one would then need to prepend to the list, and in case of multiple same issue might arise, using a mapping prevents such occurrence.